### ISSUE TITLE
Debugging and introspecting java code via dap-mode/lsp-java 

### ISSUE THREAD

**Issue by yyoncho** [clojure-emacs/cider#2423]
I am working on a new package `dap-mode` which is using [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/) which currently supports debugging java.  Here it is guide how it can be configured to debug java code when doing `cider` repl session: [debug java](https://github.com/yyoncho/dap-mode/wiki/HowTo%3A-Debug-Java-code-when-using-cider-REPL-session) . I believe that it may be usefull for clojure programmer who are doing java interoperability programming.

Additionally, if `cider` provides information for the currenly executed line when the debugger is in clj source file it may be used to debug clojure source as well.

**Comment by bbatsov** [comment]
> I am working on a new package dap-mode which is using Debug Adapter Protocol which currently supports debugging java. Here it is guide how it can be configured to debug java code when doing cider repl session: debug java . I believe that it may be usefull for clojure programmer who are doing java interoperability programming.

Looks cool! I guess we should mention this somewhere in the manual. 

> Additionally, if cider provides information for the currenly executed line when the debugger is in clj source file it may be used to debug clojure source as well.

What exactly do you need?

**Comment by bbatsov** [comment]
>  If cider provides a way to get the location in the clj file for particular stack frame then the debugger would be able to step through CLJ files as well.

That shouldn't be hard IMO. What should be the API for this?

**Comment by jumarko** [comment]
@yyoncho This looks really cool.
Do you still plan to work on this?

**Comment by yyoncho** [comment]
@jumarko yes, I am working on it, now it supports 10+ languages out of the box and it is part of emacs-lsp organization. Do you have something particular in mind?

@bbatsov I guess something like: `(thread-name, stack-frame) -> (buffer-or-file-name, point)` would be sufficient to support stepping through. 

Apart from that, I think that debugging clojurescript is possible even now via FF/Chrome debuggers which since they accept sourcemaps. 

PS: I live in Sofia so we have a beer and chat about emacs and maintaining open source projects(I am now maintaining lsp-mode as well).

**Comment by bbatsov** [comment]
> @bbatsov I guess something like: (thread-name, stack-frame) -> (buffer-or-file-name, point) would be sufficient to support stepping through.

Got it. @vspinu @Malabarba Would one of you be able to look into this? 

> PS: I live in Sofia so we have a beer and chat about emacs and maintaining open source projects(I am now maintaining lsp-mode as well).

Small world! I'd love to meet a fellow Bulgarian Emacser! Shoot me some email or something and we'll make it happen!

**Comment by vspinu** [comment]
I would love to have a look at this but unfortunately I am really having hard time figuring lsp/lsp-java machinery for now, let alone dap-java. But it's probably my problem. 

@yyoncho your original guide link is empty. Would be really great if you could provide a minimal setup for how a cider/clojure user could delve into dap. Thanks!

**Comment by yyoncho** [comment]
Here it is the proper link https://github.com/emacs-lsp/dap-mode/wiki/Clojure:-debugging-and-introspecting-Java-Code . I will create an `emacs -q` configuration.

**Comment by yyoncho** [comment]
> @yyoncho your original guide link is empty. Would be really great if you could provide a minimal setup for how a cider/clojure user could delve into dap. Thanks!

@vspinu is your issue sorted out now?

**Comment by vspinu** [comment]
Yes, thanks. I have figured out the lsp part. Things do work as expected. Nice.  Next is dap, but it might take a few weeks till I am able to dwell properly into it. I am in a traveling mode now.

**Comment by vspinu** [comment]
Bot, pls keep this open.

**Comment by ericdallo** [comment]
Hey, any news on this issue?
 it would be really cool to have this feature
